processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +130,75 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
+    final Integer priorIdex = facts.get(key);
+
+    Aggregator[] aggs;
+
+    if (null != priorIdex) {
+      aggs = atomicGet(priorIdex);
```

Comment:
technically our atomicGet/atomicSet are just thread safe not atomic, so it might be confusing to call them that?